arangod/VocBase/KeyGenerator.cpp

```diff
@@ -104,67 +111,510 @@ static std::array<bool, 256> const keyCharLookupTable = { {
+  UUID = 3,
+  PADDED = 4
+};
+
+
+/// @brief base class for traditional key generators
+class TraditionalKeyGenerator : public KeyGenerator {
+ public:
+  /// @brief create the generator
+  explicit TraditionalKeyGenerator(bool allowUserKeys)
+    : KeyGenerator(allowUserKeys) {}
 
-}
+  /// @brief return the generator name (must be lowercase)
+  static char const* name() { return "traditional"; }
+  
+  /// @brief track usage of a key - default implementation is to throw!
+  void track(char const*, size_t) override {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "this key generator cannot track keys");
+  }
+  
+  /// @brief build a VelocyPack representation of the generator in the builder
+  void toVelocyPack(arangodb::velocypack::Builder& builder) const override {
+    TRI_ASSERT(!builder.isClosed());
+    builder.add("type", VPackValue(name()));
+    builder.add("allowUserKeys", VPackValue(_allowUserKeys));
+  }
+};
 
-/// @brief create the key generator
-KeyGenerator::KeyGenerator(bool allowUserKeys)
-    : _allowUserKeys(allowUserKeys) {}
+/// @brief traditional key generator for a single server
+class TraditionalKeyGeneratorSingle final : public TraditionalKeyGenerator {
+ public:
+  /// @brief create the generator
+  explicit TraditionalKeyGeneratorSingle(bool allowUserKeys) 
+    : TraditionalKeyGenerator(allowUserKeys), _lastValue(0) {}
+
+  bool trackKeys() const override { return true; }
+
+  /// @brief generate a key
+  std::string generate() override {
+    TRI_voc_tick_t tick = TRI_NewTickServer();
+    
+    {
+      MUTEX_LOCKER(mutexLocker, _lock);
+
+      if (tick <= _lastValue) {
+        tick = ++_lastValue;
+      } else {
+        _lastValue = tick;
+      }
+    }
+
+    if (tick == UINT64_MAX) {
+      // sanity check
+      return std::string();
+    }
+    return arangodb::basics::StringUtils::itoa(tick);
+  }
+  
+  /// @brief validate a key
+  int validate(char const* p, size_t length, bool isRestore) override {
+    int res = KeyGenerator::validate(p, length, isRestore);
+
+    if (res != TRI_ERROR_NO_ERROR) {
+      return res;
+    }
+
+    track(p, length);
+
+    return TRI_ERROR_NO_ERROR;
+  }
+
+  /// @brief track usage of a key
+  void track(char const* p, size_t length) override {
+    // check the numeric key part
+    if (length > 0 && p[0] >= '0' && p[0] <= '9') {
+      // potentially numeric key
+      uint64_t value = NumberUtils::atoi_zero<uint64_t>(p, p + length);
+
+      if (value > 0) {
+        MUTEX_LOCKER(mutexLocker, _lock);
+
+        if (value > _lastValue) {
+          // and update our last value
+          _lastValue = value;
+        }
+      }
+    }
+  }
+
+  /// @brief build a VelocyPack representation of the generator in the builder
+  void toVelocyPack(arangodb::velocypack::Builder& builder) const override {
+    TRI_ASSERT(!builder.isClosed());
+    TraditionalKeyGenerator::toVelocyPack(builder);
+    builder.add("lastValue", VPackValue(_lastValue));
```

Comment:
lastValue seems to be a bit of a magic string in rocksdb, maybe it would be a good idea too put this in `StaticStrings`